File: prototypes/python/yin_psola/pyin.py
import numpy as np
from typing import Tuple

def pyin(y: np.ndarray, 
         frame_length: int = 2048,
         hop_length: int = 512,
         sr: int = 44100,
         fmin: float = 65.0,
         fmax: float = 2093.0,
         threshold: float = 0.1) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
    """
    Implements the Probabilistic YIN (PYIN) pitch detection algorithm.
    
    Args:
        y: Input audio signal
        frame_length: Size of each analysis frame
        hop_length: Number of samples between successive frames
        sr: Sample rate
        fmin: Minimum frequency to detect (Hz)
        fmax: Maximum frequency to detect (Hz)
        threshold: Threshold for pitch detection
    
    Returns:
        Tuple containing:
        - Fundamental frequency estimates (f0)
        - Voice/unvoiced flag for each frame
        - Voiced probability for each frame
    """
    
    # Calculate number of frames
    n_frames = 1 + (len(y) - frame_length) // hop_length
    
    # Initialize output arrays
    f0 = np.zeros(n_frames)
    voiced_flag = np.zeros(n_frames, dtype=bool)
    voiced_prob = np.zeros(n_frames)
    
    for i in range(n_frames):
        # Get current frame
        start = i * hop_length
        frame = y[start:start + frame_length]
        
        # Compute difference function
        diff = difference_function(frame)
        
        # Compute cumulative mean normalized difference
        cmnd = cumulative_mean_normalized_difference(diff)
        
        # Find pitch period candidates
        tau_candidates = get_pitch_period_candidates(cmnd, threshold)
        
        if len(tau_candidates) > 0:
            # Convert best period to frequency
            period = tau_candidates[0]
            if period != 0:
                freq = sr / period
                
                # Check if frequency is within bounds
                if fmin <= freq <= fmax:
                    f0[i] = freq
                    voiced_flag[i] = True
                    voiced_prob[i] = 1.0 - cmnd[period]
    
    return f0, voiced_flag, voiced_prob

def difference_function(frame: np.ndarray) -> np.ndarray:
    """
    Compute the difference function for YIN algorithm.
    """
    frame_length = len(frame)
    diff = np.zeros(frame_length)
    
    for tau in range(frame_length):
        for j in range(frame_length - tau):
            diff[tau] += (frame[j] - frame[j + tau]) ** 2
            
    return diff

def cumulative_mean_normalized_difference(diff: np.ndarray) -> np.ndarray:
    """
    Compute the cumulative mean normalized difference function.
    """
    cmnd = np.zeros_like(diff)
    cmnd[0] = 1.0
    
    running_sum = 0.0
    for tau in range(1, len(diff)):
        running_sum += diff[tau]
        cmnd[tau] = diff[tau] * tau / running_sum if running_sum > 0 else 1.0
    
    return cmnd

def get_pitch_period_candidates(cmnd: np.ndarray, threshold: float) -> np.ndarray:
    """
    Find pitch period candidates from the CMND function.
    """
    # Find all local minima
    candidates = []
    
    for i in range(1, len(cmnd) - 1):
        if cmnd[i] < cmnd[i-1] and cmnd[i] < cmnd[i+1] and cmnd[i] < threshold:
            candidates.append(i)
    
    return np.array(candidates)

File: app/src/audio/autotune/pyin.rs
use crate::audio::Key;
use ndarray::{Array1, ArrayView1, s};

#[derive(Debug, Clone)]
pub struct PYinResult {
    pub f0: Array1<f32>,
    pub voiced_flag: Array1<bool>,
    pub voiced_prob: Array1<f32>,
}
impl PYinResult {
    pub fn snap_to_scale(&self, key: Key) -> Array1<f32> {
        let scale_frequencies = key.get_scale_frequencies(2, 6); // From octave 2 to 6
        self.f0
            .iter()
            .map(|&freq| {
                if freq <= 0.0 {
                    return 0.0;
                }
                let mut closest_freq = scale_frequencies[0];
                let mut min_diff = (freq - closest_freq).abs();
                for &scale_freq in &scale_frequencies[1..] {
                    let diff = (freq - scale_freq).abs();
                    if diff < min_diff {
                        min_diff = diff;
                        closest_freq = scale_freq;
                    }
                }
                closest_freq
            })
            .collect()
    }
}

pub fn difference_function(frame: &ArrayView1<f32>) -> Array1<f32> {
    let frame_length = frame.len();
    let mut diff = Array1::zeros(frame_length);

    for tau in 1..frame_length {
        let mut sum = 0.0;
        for j in 0..(frame_length - tau) {
            let delta = frame[j] - frame[j + tau];
            sum += delta * delta;
        }
        diff[tau] = sum;
    }

    diff
}

pub fn cumulative_mean_normalized_difference(diff: &Array1<f32>) -> Array1<f32> {
    let frame_length = diff.len();
    let mut cmnd = Array1::zeros(frame_length);
    cmnd[0] = 1.0;

    let mut running_sum = 0.0;
    for tau in 1..frame_length {
        running_sum += diff[tau];
        cmnd[tau] = diff[tau] * (tau as f32) / running_sum.max(1e-10);
    }

    cmnd
}

pub fn get_pitch_period_candidates(cmnd: &Array1<f32>, threshold: f32) -> Vec<usize> {
    let mut candidates = Vec::new();
    let frame_length = cmnd.len();

    for tau in 1..frame_length - 1 {
        if cmnd[tau] < threshold && cmnd[tau] < cmnd[tau - 1] && cmnd[tau] < cmnd[tau + 1] {
            candidates.push(tau);
        }
    }

    candidates
}

pub fn pyin(
    y: &Array1<f32>,
    frame_length: usize,
    hop_length: usize,
    sample_rate: u32,
    f_min: f32,
    f_max: f32,
    threshold: f32,
) -> PYinResult {
    let n_frames = 1 + (y.len() - frame_length) / hop_length;

    let mut f0 = Array1::zeros(n_frames);
    let mut voiced_flag = Array1::from_elem(n_frames, false);
    let mut voiced_prob = Array1::zeros(n_frames);

    for i in 0..n_frames {
        let start = i * hop_length;
        let frame = y.slice(s![start..start + frame_length]);

        let diff = difference_function(&frame);
        let cmnd = cumulative_mean_normalized_difference(&diff);
        let candidates = get_pitch_period_candidates(&cmnd, threshold);

        if !candidates.is_empty() {
            let period = candidates[0];
            if period > 0 {
                let freq = sample_rate as f32 / period as f32;

                if f_min <= freq && freq <= f_max {
                    f0[i] = freq;
                    voiced_flag[i] = true;
                    voiced_prob[i] = 1.0 - cmnd[period];
                }
            }
        }
    }

    PYinResult {
        f0,
        voiced_flag,
        voiced_prob,
    }
}
